processing/src/main/java/io/druid/query/IntervalChunkingQueryRunner.java

```diff
@@ -31,57 +34,90 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ExecutorService;
 
 /**
  */
 public class IntervalChunkingQueryRunner<T> implements QueryRunner<T>
 {
   private final QueryRunner<T> baseRunner;
-  private final Period period;
 
-  public IntervalChunkingQueryRunner(QueryRunner<T> baseRunner, Period period)
+  private final QueryToolChest<T, Query<T>> toolChest;
+  private final ExecutorService executor;
+  private final QueryWatcher queryWatcher;
+  private final ServiceEmitter emitter;
+
+  public IntervalChunkingQueryRunner(QueryRunner<T> baseRunner, QueryToolChest<T, Query<T>> toolChest,
+      ExecutorService executor, QueryWatcher queryWatcher, ServiceEmitter emitter)
   {
     this.baseRunner = baseRunner;
-    this.period = period;
+    this.toolChest = toolChest;
+    this.executor = executor;
+    this.queryWatcher = queryWatcher;
+    this.emitter = emitter;
   }
 
   @Override
   public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
   {
-    if (period.getMillis() == 0) {
+    final Period chunkPeriod = getChunkPeriod(query);
+    if (chunkPeriod.toStandardDuration().getMillis() == 0) {
       return baseRunner.run(query, responseContext);
     }
 
-    return Sequences.concat(
-        FunctionalIterable
-            .create(query.getIntervals())
-            .transformCat(
-                new Function<Interval, Iterable<Interval>>()
+    List<Interval> chunkIntervals = Lists.newArrayList(FunctionalIterable
+        .create(query.getIntervals())
+        .transformCat(
+            new Function<Interval, Iterable<Interval>>()
+            {
+              @Override
+              public Iterable<Interval> apply(Interval input)
+              {
+                return splitInterval(input, chunkPeriod);
+              }
+            }
+        ));
+
+    if(chunkIntervals.size() <= 1) {
+      return baseRunner.run(query, responseContext);
+    }
+
+    final QueryRunner<T> finalQueryRunner = new AsyncQueryRunner<T>(
+        //Note: it is assumed that toolChest.mergeResults(..) gives a query runner that is
+        //not lazy i.e. it does most of its work on call to run() method 
+        toolChest.mergeResults(
+            new MetricsEmittingQueryRunner<T>(
+                emitter,
+                new Function<Query<T>, ServiceMetricEvent.Builder>()
                 {
                   @Override
-                  public Iterable<Interval> apply(Interval input)
+                  public ServiceMetricEvent.Builder apply(Query<T> input)
                   {
-                    return splitInterval(input);
+                    return toolChest.makeMetricBuilder(query);
```

Comment:
also can you check that the query/time emmited contains the correct chunked interval instead of full query interval